server/src/main/java/com/google/musicanalysis/site/OAuthLoginServlet.java

```diff
@@ -2,40 +2,62 @@
 
 import java.io.IOException;
 import java.math.BigInteger;
-import java.net.URLEncoder;
 import java.security.SecureRandom;
 import javax.servlet.ServletException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import com.google.musicanalysis.util.URLEncodedBuilder;
 
-@WebServlet("/api/oauth/login")
+@WebServlet("/api/oauth/login/*")
 public class OAuthLoginServlet extends HttpServlet {
 
   @Override
   protected void doGet(HttpServletRequest req, HttpServletResponse res)
       throws ServletException, IOException {
-    // generate a nonce to prevent CSRF
-    var state = new BigInteger(130, new SecureRandom()).toString(32);
-    req.getSession(true).setAttribute("oauth:state", state);
+    var path = req.getPathInfo().substring(1);
+
+    if (path == null) {
+      res.setStatus(404);
+      return;
+    }
 
     // URI that user should be redirected to after logging in
     var redirectUri = System.getenv().get("OAUTH_CALLBACK_URI");
+    String oauthLoginUriBase;
+    URLEncodedBuilder oauthLoginUriParams;
+
+      // generate a nonce to prevent CSRF
```

Comment:
nit: move comment back 
```suggestion
    // generate a nonce to prevent CSRF
```